### PR TITLE
EVAL-02: basic work on compiler API wrapper

### PR DESCRIPTION
**Pull request by jyoo980**
I've also included the example script from the TypeScript compiler API page.

Changing some of the values in that file with the values of our function calls should be a quick + dirty way for us to test our code other than unit testing.

### CODE REVIEW

**Review by scveloso — APPROVED** [review]
i like it i like it

File: src/ast/DirDecl.ts
Comment (uslava77) [inline src/ast/DirDecl.ts]: dirs are different now (nested version)

File: src/ast/FieldDecl.ts
Comment (uslava77) [inline src/ast/FieldDecl.ts]: should these extend an abstract class Node?

File: src/ast/DirDecl.ts
Comment (uslava77) [inline src/ast/DirDecl.ts]: should these extend an abstract class Node?
and parse evaluate functions throughout?

File: src/ast/ModuleDecl.ts
Comment (uslava77) [inline src/ast/ModuleDecl.ts]: should these extend an abstract class Node?

File: src/ast/ModuleDecl.ts
Comment (uslava77) [inline src/ast/ModuleDecl.ts]: should have quotations around modules

File: src/ast/FieldDecl.ts
Comment (uslava77) [inline src/ast/FieldDecl.ts]: this is different now

```
fields public [string bar, boolean baz]
        generate getters
        generate setters
```

File: src/ast/DirDecl.ts
Comment (uslava77) [inline src/ast/DirDecl.ts]: should initialization happen at parse()?

Comment (uslava77) [reply src/ast/DirDecl.ts]: this comment applies for all constructors in this PR

File: src/ast/VarList.ts
Comment (uslava77) [inline src/ast/VarList.ts]: [string foo, boolean bar, IInsightFacade baz]

**Review by uslava77 — COMMENTED** [review]
Haven't finished but here are some comments, will look at the other files tomorrow

**Comment by jyoo980** [comment]
@uslava77 

Addressing your comments - I made these before we finalized the EBNF so I'll make the necessary changes and push again!

File: src/ast/AstNode.ts
Comment (kiyomih) [inline src/ast/AstNode.ts:2]: nit: would prefer if this is an interface unless we have a specific reason for making this an abstract class (like we're planning to include some implementation here). 
Reasons: we can't easily extend multiple classes in typescript (and there may be a case we want to do this at some point?), and if we're just implementing this abstract class, then it's just treated as an interface anyways.

**Review by kiyomih — APPROVED** [review]
Looks great! 🔥💯🙌🏻 Just left a small nit comment, but would be good with this being merged as is.

File: src/ast/AstNode.ts
Comment (jyoo980) [reply src/ast/AstNode.ts:2]: @kiyomih thanks for your feedback!

I agree that usually, behaviour is better specified as an interface. I was thinking that we have some of the common classes, e.g. `Tokenizer`, `FileSystem`, live inside the abstract `AstNode` class, much like how Elisa implemented tinyDot in class. 

I'll wait for your response before I merge; architectural designs like these are important, thanks for bringing it up!

Comment (kiyomih) [reply src/ast/AstNode.ts:2]: That's a fair point! Everything looks good to me in that case.

File: src/ast/symbols/TypeTable.ts
Comment (daviidli) [inline src/ast/symbols/TypeTable.ts]: Is this returning true iff **all** candidates are valid types?
If so, we can do this instead: 
```typescript
candidates.every((candidate) => this.isValidType(candidate));
```
As is, it just returns if the last candidate is valid or not

**Review by daviidli — APPROVED** [review]
Looks good to me, just a small comment.

File: src/ast/symbols/TypeTable.ts
Comment (jyoo980) [reply src/ast/symbols/TypeTable.ts]: You're right - changing!

**Pull request merged**
